images/python/main.tf

```diff
@@ -1,29 +1,73 @@
 terraform {
   required_providers {
     apko = { source = "chainguard-dev/apko" }
+    oci  = { source = "chainguard-dev/oci" }
   }
 }
 
 variable "target_repository" {
   description = "The docker repo into which the image and attestations should be published."
 }
 
+locals {
+  # TODO: we don't have a virtual package that we can use for PIP (below)
+  # so we still use explicit version streams for this.
+  version = "3.11"
+}
+
+module "latest-config" {
+  source = "./config"
+
+  extra_packages = ["python-${local.version}"]
+}
+
+module "latest" {
+  source = "../../tflib/publisher"
+
+  name = basename(path.module)
+
+  target_repository = var.target_repository
+  config            = module.latest-config.config
+}
+
 module "dev" { source = "../../tflib/dev-subvariant" }
 
-module "tagger" {
-  source = "../../tflib/tagger"
-
-  depends_on = [
-    module.test-three-ten,
-    module.test-three-ten-dev,
-    module.test-three-eleven,
-    module.test-three-eleven-dev,
-  ]
-
-  tags = merge(
-    { for t in toset(module.version-tags-3_10.tag_list) : t => module.three-ten.image_ref },
-    { for t in toset(module.version-tags-3_10.tag_list) : "${t}-dev" => module.three-ten-dev.image_ref },
-    { for t in toset(concat(["latest"], module.version-tags-3_11.tag_list)) : t => module.three-eleven.image_ref },
-    { for t in toset(concat(["latest"], module.version-tags-3_11.tag_list)) : "${t}-dev" => module.three-eleven-dev.image_ref },
-  )
+module "latest-dev" {
+  source = "../../tflib/publisher"
+
+  name = basename(path.module)
+
+  target_repository = var.target_repository
+  # Make the dev variant an explicit extension of the
+  # locked original.
+  config = jsonencode(module.latest.config)
+  extra_packages = concat(module.dev.extra_packages, [
+    "build-base",
+    "py${local.version}-pip",
+    "python-${local.version}-dev"
```

Comment:
We [have](https://github.com/wolfi-dev/os/blob/617c5cb8acc5de499d9178f12c88bd117043c3db/python-3.11.yaml#L10-L11) a `python-3` virtual, and we [have](https://github.com/wolfi-dev/os/blob/617c5cb8acc5de499d9178f12c88bd117043c3db/python-3.11.yaml#L102) a `python3-dev` virtual.

However, we do not have one for `py3-pip` which we would need to make our public images strictly track virtuals, so we can get rid of the explicit `3.11` we have checked in here and rely exclusively on the APK version resolution logic.

Is this as simple as creating a `py3-pip` virtual that these pip packages provide?

Reply:
Bah, we do have `py3-pip`!

Reply:
https://github.com/wolfi-dev/os/blob/617c5cb8acc5de499d9178f12c88bd117043c3db/py3.11-pip.yaml#L11C9-L11C16

Reply:
I think I missed this because we are pretty inconsistent about applying `provides` to older versions of things, e.g. `py3.10-pip` doesn't have a `provides` for `py3-pip`, which may have been what I was looking at.

Reply:
cool that seems to pass.